pubsub/pubsub.go

```diff
@@ -0,0 +1,272 @@
+// tune how messages are sent. Behind the scenes, NewTopic spins up a goroutine
+// to bundle messages into batches and send them to the server.
+func NewTopic(ctx context.Context, d driver.Topic, opts TopicOptions) *Topic {
+	handler := func(item interface{}) {
+		mecs, ok := item.([]msgErrChan)
+		if !ok {
+			panic("failed conversion to []msgErrChan in bundler handler")
+		}
+		var dms []*driver.Message
+		for _, mec := range mecs {
+			m := mec.msg
+			dm := &driver.Message{
+				Body:       m.Body,
+				Attributes: m.Attributes,
+				AckID:      m.ackID,
+			}
+			dms = append(dms, dm)
+		}
+		err := d.SendBatch(ctx, dms)
+		for _, mec := range mecs {
+			mec.errChan <- err
+		}
+	}
+	b := bundler.NewBundler(msgErrChan{}, handler)
+	if opts.SendDelay == 0 {
+		b.DelayThreshold = SendDelayDefault
+	} else {
+		b.DelayThreshold = opts.SendDelay
+	}
+	if opts.BatchSize != 0 {
+		b.BundleCountThreshold = opts.BatchSize
+	}
+	t := &Topic{
+		driver:  d,
+		batcher: b,
+	}
+	return t
+}
+
+// Subscription receives published messages.
+type Subscription struct {
+	driver driver.Subscription
+
+	// ackBatcher makes batches of acks and sends them to the server.
+	ackBatcher *bundler.Bundler
+
+	// q is the local queue of messages downloaded from the server.
+	sem chan struct{}
+	q   []*Message
+}
+
+// SubscriptionOptions contains configuration for Subscriptions.
+type SubscriptionOptions struct {
+	// AckDelay tells the max duration to wait before sending the next batch
+	// of acknowledgements back to the server.
+	AckDelay time.Duration
+
+	// AckBatchSize is the maximum number of acks that should be sent to
+	// the server in a batch.
+	AckBatchSize int
+
+	// AckDeadline tells how long the server should wait before assuming a
+	// received message has failed to be processed.
+	AckDeadline time.Duration
+}
+
+// AckDelayDefault is the value for SubscriptionOptions.AckDelay if it is not set.
+const AckDelayDefault = time.Millisecond
+
+// Receive receives and returns the next message from the Subscription's queue,
+// blocking and polling if none are available. This method can be called
+// concurrently from multiple goroutines. The Ack() method of the returned
+// Message has to be called once the message has been processed, to prevent it
+// from being received again.
+func (s *Subscription) Receive(ctx context.Context) (*Message, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-s.sem:
+	}
+	defer func() {
+		s.sem <- struct{}{}
+	}()
+	if len(s.q) == 0 {
+		if err := s.getNextBatch(ctx); err != nil {
+			return nil, err
+		}
+	}
+	m := s.q[0]
+	s.q = s.q[1:]
+	return m, nil
+}
+
+// getNextBatch gets the next batch of messages from the server.
+func (s *Subscription) getNextBatch(ctx context.Context) error {
+	msgs, err := s.driver.ReceiveBatch(ctx)
+	if err != nil {
+		return err
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		if len(msgs) == 0 {
```

Comment:
Maybe this should be allowed? E.g., some impls may do polling?
If not, make sure to document the requirement in the driver package.

Reply:
@jba was saying it should not be, since different drivers may have different ways to efficiently block until a message is available. We can revisit this later if we find that some drivers need to be able to return an empty message slice.

Reply:
If the driver wants to do polling, that's its problem. The concrete impl should never have to poll, though.